remove iris from test suite

dodgejoel opened the pull request:
- adds inlined toy dataset for e2e test
- adds disabled test demonstrating functionality to be changed

Given my lack of experience working with git hub workflows, I am going to take it slow getting these changes in. I believe this single commit PR is relatively modest in scope. Let me know if you had something else in mind for the test data to replace iris.

Commit:
remove iris from test suite

- adds inlined toy dataset for e2e test
- adds disabled test demonstrating functionality to be changed

hlin117 commented on tests/test_mdlp.py:
```diff
@@ -1,3 +1,5 @@
+import itertools
+import pytest
```
Let's alphabetize this.

dodgejoel replied:
👍

hlin117 commented on tests/test_mdlp.py:
```diff
@@ -32,165 +34,52 @@ def test_find_cut_no_cut():
-def test_mdlp_iris():
-    from sklearn.datasets import load_iris
+def test_e2e():
+  X = np.array([
```
You actually don't need `np.array(...)`, because `fit_transform` will transform the input into a `np.array`.

dodgejoel replied:
This is empirically false.

hlin117 replied:
What do you mean?

And if it's not the case, then we have a bug, and we should look into that.

dodgejoel replied:
What I mean is that if I attempt to call fit_transform on these data without casting them to an np.array first then I get an AttributeError from L110 in discretization.py where we are trying to get the shape of X.

BaseEstimator.fit_transform docstring suggests that X and y are meant to be np.arrays already and does not do any conversion. I don't see a bug here I just think that the arguments to fit_transform are supposed to be np.array.

Is it possible that this has changed in sklearn between versions or something? How sure are you that a cast was done at some point?

dodgejoel replied:
Sorry I misspoke I meant TransformerMixin.fit_transform of course 😝

hlin117 replied:
See #30 . My mistake!

hlin117 commented on tests/test_mdlp.py:
```diff
@@ -32,165 +34,52 @@ def test_find_cut_no_cut():
 
-def test_mdlp_iris():
-    from sklearn.datasets import load_iris
+def test_e2e():
```
Let's rename this to `test_fit_transform`

dodgejoel replied:
👍

hlin117 commented on tests/test_mdlp.py:
```diff
@@ -32,165 +34,52 @@ def test_find_cut_no_cut():
+  y = np.array([0, 0, 0, 1])
+  transformed = MDLP(shuffle=False).fit_transform(X, y)
+  expected = np.array([
+    [0, 0],
```
This is a bit of a precarious test case. Everything seems to discretize to zero. Can you come up with a test case with a more variety of outputs?

dodgejoel replied:
The two columns display both successful and unsuccessful discretization of data. It seems valuable to have a failure case here that discretizes to all 0s. The first column is not all 0 in this test case already but I will modify to make it even more interesting.

hlin117 commented on tests/test_mdlp.py:
```diff
@@ -32,165 +34,52 @@ def test_find_cut_no_cut():
-    from sklearn.datasets import load_iris
+def test_e2e():
+  X = np.array([
+    [0.1, 0.1],
```
So instead of having decimal values here, why not multiply each value by 10? Discretization should be invariant under scaling.

dodgejoel replied:
I think that having decimals for the initial test data is reasonable as a way to emphasize that discretization operates on continuous data. The invariance under scaling is interesting though. I will add several assertions demonstrating the invariance of discretization under several transformations of the data.

hlin117 commented on tests/test_mdlp.py:
```diff
@@ -32,165 +34,52 @@ def test_find_cut_no_cut():
-
-    assert_array_equal(transformed, expected,
-                       err_msg="MDLP output is inconsistent with previous runs.")
+@pytest.mark.skip(reason='demonstrating behavior to be fixed')
```
For now, let's defer this to a later PR.

dodgejoel replied:
👍

hlin117 commented on tests/test_mdlp.py:
```diff
@@ -32,165 +34,52 @@ def test_find_cut_no_cut():
+    [0.1, 0.1],
+    [0.2, 1.1],
+    [0.3, 0.2],
+    [1.5, .9]
```
You can also change 1.5 -> 0.4 correct?

dodgejoel replied:
Indeed. That is true. I will change the data here to remove some small red herrings as to what is going on.

hlin117 reviewed (changes requested):
It's looking good so far! Thank you for your work. I'd like to see the comments addressed though.

Commit:
further cleanup of test cases

- simplify data used in test of fit_transform
- add assertions for various data transforms that leave discretization
  invariant

dodgejoel commented:
All these suggestions make good sense to me. Thanks!

Commit:
fix comment on test case

dodgejoel commented:
How does this look?

hlin117 commented on tests/test_mdlp.py:
```diff
@@ -32,165 +34,31 @@ def test_find_cut_no_cut():
-def test_mdlp_iris():
-    from sklearn.datasets import load_iris
+def test_fit_transform():
+  X = np.array([
```
Can you rebase over my most recent commit? You should be able to pass in nested lists now.

dodgejoel replied:
i did so and added a test case for passing in raw arrays. This even uncovered some unfortunate existing behavior where integer arrays were not being cast to np.float64 and this was relevant for the resulting discretization. I changed the check_array to also have a dtype for the data.

hlin117 commented on tests/test_mdlp.py:
```diff
@@ -32,165 +34,31 @@ def test_find_cut_no_cut():
+    [1, 0],
+    [2, 0],
+  ]
+  assert_array_equal(transformed, expected)
```
For testing APIs, the `expected` variable generally comes before the `actual` value.

dodgejoel replied:
ok. what does 'generally' mean?

hlin117 commented on tests/test_mdlp.py:
```diff
@@ -32,165 +34,31 @@ def test_find_cut_no_cut():
-    y = iris.target
-    mdlp = MDLP(shuffle=False)
-    transformed = mdlp.fit_transform(X, y)
+  # discretization is invariant under rescaling of the data
```
I commend you to test things like this. But 

1. I would encourage you to put it in another test file
2. I would try out different factors 

See https://github.com/scikit-learn/scikit-learn/blob/discrete/sklearn/preprocessing/tests/test_discretization.py#L173.

hlin117 commented on tests/test_mdlp.py:
```diff
@@ -32,165 +34,31 @@ def test_find_cut_no_cut():
-        [ 2.,  1.,  2.,  2.],
-        [ 2.,  0.,  2.,  2.]]
+  # discretization is invariant under translation of the data
+  translated_disc = MDLP(shuffle=False).fit_transform(X - 5, y)
```
This is nice to test! Can you put it in another test case though?

dodgejoel replied:
yup.

hlin117 commented on tests/test_mdlp.py:
```diff
@@ -32,165 +34,31 @@ def test_find_cut_no_cut():
-    assert_array_equal(transformed, expected,
-                       err_msg="MDLP output is inconsistent with previous runs.")
+  # discretization is invariant under renaming of the labels
+  relabeled_disc = MDLP(shuffle=False).fit_transform(X, y + 1)
```
I think we can safely remove this test case.

dodgejoel replied:
ok.

hlin117 reviewed (changes requested):
Thanks again for all your hard work. Can you take a look at my comments though? This is close to being merged, and I'm certainly looking forward to it.

Commit:
re-order expected <> transformed in test_fit_transform

Commit:
split unit tests - modify cast to array by adding type

separate unit tests covering
- invariance of discretization under rescaling
- invariance of discretization under translation
- coercion to np.array of list args

Commit:
rm pytest import in test_mdlp

dodgejoel commented:
I have made changes which I think respond to all of your comments / requests. LMK if you had something else in mind.

hlin117 commented on tests/test_mdlp.py:
```diff
@@ -32,165 +31,46 @@ def test_find_cut_no_cut():
-    y = iris.target
-    mdlp = MDLP(shuffle=False)
-    transformed = mdlp.fit_transform(X, y)
+def test_coerce_list():
```
I appreciate the thought of adding this test, because I previously did not test whether we can use numpy arrays versus lists.

Thanks. If this were the standard scikit-learn library, there would already be unit tests for this functionality.

hlin117 commented on tests/test_mdlp.py:
```diff
@@ -32,165 +31,46 @@ def test_find_cut_no_cut():
 
-def test_mdlp_iris():
-    from sklearn.datasets import load_iris
+def test_fit_transform():
```
I think you should call this

```
def test_fit_transform_scale
```

hlin117 commented on tests/test_mdlp.py:
```diff
@@ -32,165 +31,46 @@ def test_find_cut_no_cut():
+    scaled_disc = MDLP(shuffle=False).fit_transform(X / 10**i, y)
+    assert_array_equal(expected, scaled_disc)
+
+def test_translation():
```
How about

```
def test_fit_transform_translate`
```

hlin117 reviewed (approved):
Thank you so much for these changes, @dodgejoel . After you update with my changes, I'll definitely merge this PR.

Commit:
rename test methods

dodgejoel commented:
All fine suggestions and they are done. Thanks @hlin117!

hlin117 commented:
Thanks for the PR, @dodgejoel ! 🥂
